### PR TITLE
Feature/10349 crypto functions part1/10433 check certificate chain

### PR DESCRIPTION
**Pull request by ArturFriesen**
## Description
Adds function to check a certificate agains a list of JKWSets. Please see tech spec (expand the diff): https://github.com/corona-warn-app/cwa-app-tech-spec/pull/110/files#diff-9d53e1eda3d6bef7bbc4898a6979ba45a08fb5eacac575ea1c5efabb8a6ad187R367-R411

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10433

### CODE REVIEW

File: src/xcode/ENA/ENASecurity/Package.swift [hunk 1]
```diff
@@ -17,13 +17,14 @@ let package = Package(
     dependencies: [
         // Dependencies declare other packages that this package depends on.
         // .package(url: /* package url */, from: "1.0.0"),
+        .package(url: "https://github.com/filom/ASN1Decoder", .upToNextMajor(from: "1.8.0"))
```
Comment (marcussc) [inline src/xcode/ENA/ENASecurity/Package.swift:20]: Please update the licences file and our generated view model for the licences

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import Foundation
+import CommonCrypto
+
+extension Data {
+
+    func sha256() -> Data {
+        // via https://www.agnosticdev.com/content/how-use-commoncrypto-apis-swift-5
+
+        // #define CC_SHA256_DIGEST_LENGTH     32
```
Comment (marcussc) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift:13]: Is this needed?

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        // Creates an array of unsigned 8 bit integers that contains 32 zeros
+        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
+
+        // CC_SHA256 performs digest calculation and places the result in the caller-supplied buffer for digest (md)
+        // Takes the strData referenced value (const unsigned char *d) and hashes it into a reference to the digest parameter.
+        _ = self.withUnsafeBytes {
+            // CommonCrypto
+            // extern unsigned char *CC_SHA256(const void *data, CC_LONG len, unsigned char *md)  -|
```
Comment (marcussc) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift:21]: Hö?

File: src/xcode/ENA/ENASecurity/Package.swift (hunk 1 above)
Comment (marcussc) [reply src/xcode/ENA/ENASecurity/Package.swift:20]: Nvm just noticed we already use it :D

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift (hunk 2 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift]: Removed

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift [hunk 4]
```diff
@@ -0,0 +1,32 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+import CommonCrypto
+
+extension Data {
+
+    func sha256() -> Data {
```
Comment (nickguendling) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/Utils/Data+SHA256.swift:10]: Could also be
```suggestion
    var sha256: Data {
```

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/TrustEvaluation.swift [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    case CERT_CHAIN_EMTPY
+    case CERT_PIN_NO_JWK_FOR_KID
+    case CERT_PIN_MISMATCH
+}
+
+class TrustEvaluation {
+
+    func check(trust: SecTrust, against jwkSet: [Data]) -> Result<Void, TrustEvaluationError> {
+        // Extract leafCertificate: the leafCertificate shall be extracted from the certificateChain. This is typically the first certificate of the chain.
+        if let serverCertificate = SecTrustGetCertificateAtIndex(trust, SecTrustGetCertificateCount(trust) - 1),
```
Comment (nickguendling) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/TrustEvaluation.swift:17]: The comment says "first certificate in the chain", but seems like the code takes the last?

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/TrustEvaluation.swift [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+           let serverPublicKey = SecCertificateCopyKey(serverCertificate),
+           let serverPublicKeyData = SecKeyCopyExternalRepresentation(serverPublicKey, nil ) as Data? {
```
Comment (nickguendling) [inline src/xcode/ENA/ENASecurity/Sources/ENASecurity/TrustEvaluation.swift:19]:
```suggestion
           let serverPublicKeyData = SecKeyCopyExternalRepresentation(serverPublicKey, nil) as Data? {
```

**Review by nickguendling — APPROVED** [review]

**Review by marcussc — APPROVED** [review]

**Review by naveeddotio — APPROVED** [review]

**Review by naveeddotio — APPROVED** [review]

File: src/xcode/ENA/ENASecurity/Sources/ENASecurity/TrustEvaluation.swift (hunk 5 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENASecurity/Sources/ENASecurity/TrustEvaluation.swift:17]: Good point. The comment is from the tech spec. 
I have copied the code from our certificate pinning.
Thats why I assume that the last one should be the leaf certificate in our case.
I think we will need to figure this out in a manual test.

**Pull request merged**